### PR TITLE
Aspects: fix filter by tags

### PR DESCRIPTION
**Pull request by annyhe**

### CODE REVIEW

File: api/v1/controllers/aspects.js [hunk 1]
```diff
@@ -29,6 +29,62 @@ const doPost = require('../helpers/verbs/doPost');
...
+function checkDuplicates(tagsArr) {
+  const LEN = tagsArr.length - ONE;
+
+  // store lowercase copies
+  const copyArr = [];
+  let toAdd;
+  for (let i = LEN; i >= ZERO; i--) {
+    let string = tagsArr[i];
+
+    // if the string begins with -, use the rest of the string for omparison
```
Comment (iamigo) [inline api/v1/controllers/aspects.js]: "comparison"

File: api/v1/controllers/aspects.js [hunk 2]
```diff
@@ -29,6 +29,62 @@ const doPost = require('../helpers/verbs/doPost');
 const doPut = require('../helpers/verbs/doPut');
 const u = require('../helpers/verbs/utils');
 const httpStatus = require('../constants').httpStatus;
+const ZERO = 0;
+const ONE = 1;
+/**
```
Comment (iamigo) [inline api/v1/controllers/aspects.js]: empty line before

File: api/v1/controllers/aspects.js [hunk 3]
```diff
@@ -29,6 +29,62 @@ const doPost = require('../helpers/verbs/doPost');
...
+ * Given an array, return true if there
+ * are duplicates. False otherwise.
+ *
+ * @param {Array} tagsArr The input array
+ * @returns {Boolean} whether input array
+ * contains duplicates
+ */
+function checkDuplicates(tagsArr) {
```
Comment (iamigo) [inline api/v1/controllers/aspects.js]: when a fn returns a boolean, it's good practice for the function name to reflect that, e.g. "hasDuplicates(...)" returns true if there are duplicates--it just makes it clearer what to expect

File: api/v1/controllers/aspects.js [hunk 4]
```diff
@@ -29,6 +29,62 @@ const doPost = require('../helpers/verbs/doPost');
...
+  return false;
+}
+
+/**
+ * Validates the given fields from request body or url.
+ * If fails, throws a corresponding error.
+ * @param {Object} requestBody Fields from request body
+ * @param {Object} params Fields from url
+ */
+function validateRequest(requestBody, params) {
```
Comment (iamigo) [inline api/v1/controllers/aspects.js]: This fn looks like it's validating the tags, so maybe call it "validateTags"?

File: api/v1/controllers/utils.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+
```
Comment (iamigo) [inline api/v1/controllers/utils.js]: Is this a new file? Needs copyright header

File: api/v1/swagger.yaml [hunk 6]
```diff
@@ -196,9 +196,11 @@ paths:
              case-insensitive. For example, ?tags=FOO,BAR will only return
              subjects with tags FOO or BAR. Prefix each of the tag name with
              a negative sign to indicate that a subject with that tag should
-             be excluded. For example, ?tags=-BAZ,-FOO will return only the
-             subjects with tag name not equal to BAZ or FOO. Subjects without
-             tags are not included in the result set.
+             be excluded. For example, ?tags=-BAZ,FOO will return only the
```
Comment (iamigo) [inline api/v1/swagger.yaml]: Can you check with @harshkothari410 and @shriramshankar on the wording here?

File: api/v1/controllers/utils.js (hunk 5 above)
Comment (annyhe) [inline api/v1/controllers/utils.js]: updated with header

File: api/v1/swagger.yaml [hunk 7]
```diff
@@ -184,6 +184,23 @@ paths:
...
+          in: query
+          items:
+            type: string
+            maxLength: 60
+            pattern: ^[0-9A-Za-z_][0-9A-Za-z_\\-]{1,59}$
+          description: >-
+             Comma-separated list of tags to include/exclude. Tag names are
+             case-insensitive. For example, ?tags=FOO,BAR will only return
+             subjects with tags FOO or BAR. Prefix each of the tag name with
+             a negative sign to indicate that a subject with that tag should
```
Comment (annyhe) [inline api/v1/swagger.yaml]: added towards the bottom.

**Comment by pallavi2209** [comment]
Tests failing?

**Comment by annyhe** [comment]
Yeah tests failing due to lint errors. Will fix with a new commit.

**Comment by annyhe** [comment]
@pallavi2209 Fixed the lint errors. Please review.

**Comment by annyhe** [comment]
These tests are unrelated to the changes and seem to be flapping. I will run them again.

**Comment by annyhe** [comment]
Tests passed. Please review @pallavi2209 @iamigo

File: api/v1/swagger.yaml [hunk 8]
```diff
@@ -184,6 +184,25 @@ paths:
...
+          name: tags
+          in: query
+          items:
+            type: string
+            maxLength: 60
+            pattern: ^[0-9A-Za-z_][0-9A-Za-z_\\-]{1,59}$
+          description: >-
+             Comma-separated list of tags to include/exclude. Tag names are
+             case-insensitive. For example, ?tags=FOO,BAR will only return
+             subjects with tags FOO or BAR. Prefix each of the tag name with
```
Comment (pallavi2209) [inline api/v1/swagger.yaml]: aspects* with tags

File: api/v1/swagger.yaml [hunk 9]
```diff
@@ -184,6 +184,25 @@ paths:
...
+             a negative sign to indicate that a subject with that tag should
+             be excluded. For example, ?tags=-BAZ,FOO will return only the
+             subjects with tag name not equal to BAZ or FOO. A mix of tag
```
Comment (pallavi2209) [inline api/v1/swagger.yaml]: aspects*

File: api/v1/controllers/utils.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
+/**
+ * Copyright (c) 2016, salesforce.com, inc.
```
Comment (pallavi2209) [inline api/v1/controllers/utils.js]: 2017

File: api/v1/controllers/aspects.js [hunk 11]
```diff
@@ -29,6 +30,31 @@ const doPost = require('../helpers/verbs/doPost');
...
+const ONE = 1;
+
+/**
+ * Validates the given fields from request body or url.
+ * If fails, throws a corresponding error.
+ * @param {Object} requestBody Fields from request body
+ * @param {Object} params Fields from url
+ */
+function validateTags(requestBody, params) {
+  let absolutePath = '';
```
Comment (pallavi2209) [inline api/v1/controllers/aspects.js]: Do we need this variable?

File: api/v1/controllers/utils.js [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+  const LEN = tagsArr.length - ONE;
+
+  // store lowercase copies
+  const copyArr = [];
+  let toAdd;
+  for (let i = LEN; i >= ZERO; i--) {
+    let string = tagsArr[i];
+
+    // if the string begins with -, use the rest of the string for comparison
+    toAdd = string[0] === '-' ? string.slice(1).toLowerCase() :
```
Comment (pallavi2209) [inline api/v1/controllers/utils.js]: I don't see this logic in subject tags. Is this consistent with subject tags? If no, do we need this in subject tags as well?

File: tests/api/v1/aspects/get.js [hunk 13]
```diff
@@ -80,7 +81,88 @@ describe(`api: GET ${path}`, () => {
...
+        }
+
+        expect(res.body.errors[0].type).to.equal('DuplicateFieldError');
+        done();
+      });
+    });
+  });
+
   describe('Single Values: ', () => {
+    it('filter by EXCLUDE tags returns expected values', (done) => {
```
Comment (pallavi2209) [inline tests/api/v1/aspects/get.js]: do we need these tests for multiple values as well? And combination of include,exclude

File: tests/api/v1/subjects/get.js [hunk 14]
```diff
@@ -104,7 +103,7 @@ describe(`api: GET ${path}`, () => {
       });
     });
 
-    it('GET with tag INCLUDE filter :: duplicate tags pass', (done) => {
+    it('GET with tag INCLUDE filter', (done) => {
```
Comment (pallavi2209) [inline tests/api/v1/subjects/get.js]: Nice!

**Comment by iamigo** [comment]
I think there are one or two (minor) outstanding changes requested by Pallavi... I am ready to approve after that.

File: api/v1/controllers/utils.js (hunk 12 above)
Comment (annyhe) [inline api/v1/controllers/utils.js]: This is consistent with subject tags: line 77 of api/v1/controllers/subjects.js

File: api/v1/controllers/aspects.js (hunk 11 above)
Comment (annyhe) [inline api/v1/controllers/aspects.js]: nope. I took it out.

**Comment by annyhe** [comment]
@iamigo @pallavi2209 Updated with changes

**Comment by annyhe** [comment]
Please review and comment @iamigo

**Pull request merged**